### ISSUE TITLE
DataTable: Sort on ListTableModel fails with NPE

### ISSUE THREAD

**Issue by robtha** [primefaces/primefaces#6913]
DataTable Sort on ListDataModel fails with NPE - WrappedData List lost after sort order has been changed.
Using List instead of ListDataModel works.

**Environment:**
 - PF Version: _10.0.0-SNAPSHOT_
 - JSF + version: _Mojarra 2.3.14_
 - Java Runtime: 11.0.9.1

**Example Bean**
```java
@Named
@ViewScoped
public class TestView implements Serializable {

    private DataService dataService;
    private DataModel dataModel;

    public DataModel getDataModel() {
         if ( dataModel == null ) {
              dataModel = new ListDataModel<>( dataService.getData() );
         }
         return dataModel;
    }
}
```

**Stacktrace**
```
10:06:24,955 INFO  [javax.enterprise.resource.webcontainer.jsf.context] (default task-7) java.lang.NullPointerException: java.lang.NullPointerException
	at deployment.test-SNAPSHOT.war//org.primefaces.component.datatable.feature.SortFeature.sort(SortFeature.java:154)
	at deployment.test-SNAPSHOT.war//org.primefaces.component.datatable.feature.SortFeature.decode(SortFeature.java:107)
	at deployment.test-SNAPSHOT.war//org.primefaces.component.datatable.DataTableRenderer.decode(DataTableRenderer.java:72)
	at deployment.test-SNAPSHOT.war//javax.faces.component.UIComponentBase.decode(UIComponentBase.java:509)
	at deployment.test-SNAPSHOT.war//org.primefaces.component.api.UIData.processDecodes(UIData.java:117)
	at deployment.test-SNAPSHOT.war//com.sun.faces.context.PartialViewContextImpl$PhaseAwareVisitCallback.visit(PartialViewContextImpl.java:606)
	at deployment.test-SNAPSHOT.war//com.sun.faces.component.visit.PartialVisitContext.invokeVisitCallback(PartialVisitContext.java:159)
	at deployment.test-SNAPSHOT.war//org.primefaces.component.api.UIData.visitTree(UIData.java:683)
	at deployment.test-SNAPSHOT.war//javax.faces.component.UIComponent.visitTree(UIComponent.java:1468)
	at deployment.test-SNAPSHOT.war//javax.faces.component.UIComponent.visitTree(UIComponent.java:1468)
	at deployment.test-SNAPSHOT.war//javax.faces.component.UIComponent.visitTree(UIComponent.java:1468)
	at deployment.test-SNAPSHOT.war//javax.faces.component.UIForm.visitTree(UIForm.java:355)
	at deployment.test-SNAPSHOT.war//javax.faces.component.UIComponent.visitTree(UIComponent.java:1468)
	at deployment.test-SNAPSHOT.war//javax.faces.component.UIComponent.visitTree(UIComponent.java:1468)
	at deployment.test-SNAPSHOT.war//com.sun.faces.context.PartialViewContextImpl.processComponents(PartialViewContextImpl.java:400)
	at deployment.test-SNAPSHOT.war//com.sun.faces.context.PartialViewContextImpl.processPartial(PartialViewContextImpl.java:261)
	at deployment.test-SNAPSHOT.war//org.primefaces.context.PrimePartialViewContext.processPartial(PrimePartialViewContext.java:65)
	at deployment.test-SNAPSHOT.war//javax.faces.component.UIViewRoot.processDecodes(UIViewRoot.java:1008)
	at deployment.test-SNAPSHOT.war//com.sun.faces.lifecycle.ApplyRequestValuesPhase.execute(ApplyRequestValuesPhase.java:55)
	at deployment.test-SNAPSHOT.war//com.sun.faces.lifecycle.Phase.doPhase(Phase.java:76)
	at deployment.test-SNAPSHOT.war//com.sun.faces.lifecycle.LifecycleImpl.execute(LifecycleImpl.java:177)
	at deployment.test-SNAPSHOT.war//javax.faces.webapp.FacesServlet.executeLifecyle(FacesServlet.java:707)
	at deployment.test-SNAPSHOT.war//javax.faces.webapp.FacesServlet.service(FacesServlet.java:451)
```

```XHTML
<p:dataTable id="datatable" value="#{dataTable018.dataModel}" var="lang" paginator="true" rows="3"
                         filteredValue="#{dataTable018.filteredProgLanguages}" filterEvent="enter" rowKey="#{lang.id}" selectionMode="single"
                         paginatorTemplate="{CurrentPageReport} {FirstPageLink} {PreviousPageLink} {PageLinks} {NextPageLink} {LastPageLink} {RowsPerPageDropdown}"
                         rowsPerPageTemplate="2,3,5,10">
    <p:column headerText="ID" sortBy="#{lang.id}">
        <h:outputText value="#{lang.id}"/>
    </p:column>
    <p:column headerText="Name" sortBy="#{lang.name}" filterBy="#{lang.name}" filterMatchMode="contains">
        <h:outputText value="#{lang.name}"/>
    </p:column>
    <p:column headerText="First appeared" sortBy="#{lang.firstAppeared}" filterBy="#{lang.firstAppeared}" filterMatchMode="gte">
        <h:outputText value="#{lang.firstAppeared}"/>
    </p:column>
    <p:column headerText="OS" field="mainOs.nameShort"/>
</p:dataTable>
```

**Comment by melloware** [comment]
@robtha I just wrote and Integration test using `ListDataModel` and it pages, sorts, and filters the table and all of the tests are passing against `10.0.0-SNAPSHOT`.  See the test here: https://github.com/primefaces-extensions/primefaces-integration-tests/commit/0ef2a61462a42ef42e6ad9dd367d5e7655ca048d

Can you create a reproducer for your issue because I can't seem to get an NPE.

**Comment by robtha** [comment]
@melloware  - thanks for your quick response. I am not very familiar with selenium tests but managed to modify your test to reproduce the problem.

The NPE occurs it the column references to a nested property and attribute _selectionMode_ is set to 'single'.

I have attached a patch for DataTableTest018 - hope this helps.

kind regards
robert

**Comment by melloware** [comment]
@robtha I don't see your patch attached here?  You may have to zip it first and attach the zip here.

**Comment by robtha** [comment]
@melloware sorry - did not recognize the upload warning

[DataTable_Sort_NPE_patch.txt](https://github.com/primefaces/primefaces/files/5899771/DataTable_Sort_NPE_patch.txt)

**Comment by melloware** [comment]
@robtha thank you.  @Rapster I can confirm this issue using the patch above and its a NULL being thrown inside the Lamba so the stack track is useless.

```java
list.sort((o1, o2) -> {
            for (SortMeta sortMeta : sortBy.values()) {
                comparisonResult.set(0);

                if (sortMeta.getComponent() instanceof HeaderRow) {
                    int result = compare(context, var, sortMeta, o1, o2, collator, locale);
                    comparisonResult.set(result);
                }
                else {
                    // Currently ColumnGrouping supports ui:repeat, therefore we have to use a callback
                    // and can't use sortMeta.getComponent()
                    // Later when we refactored ColumnGrouping, we may remove #invokeOnColumn as we dont support ui:repeat in other cases
                    table.invokeOnColumn(sortMeta.getColumnKey(), column -> {
                        if (column instanceof DynamicColumn) {
                            ((DynamicColumn) column).applyStatelessModel();
                        }

                        int result = compare(context, var, sortMeta, o1, o2, collator, locale);
                        comparisonResult.set(result);
                    });
                }

                if (comparisonResult.get() != 0) {
                    return comparisonResult.get();
                }
            }

            return 0;
        });
```

**Comment by Rapster** [comment]
Is this a regression? Indeed it only happens if selection enabled...

**Comment by tandraschko** [comment]
can someone provide a primefaces-test?

**Comment by melloware** [comment]
I can get it to you today @tandraschko

**Comment by Rapster** [comment]
[primefaces-test-master.zip](https://github.com/primefaces/primefaces/files/5911126/primefaces-test-master.zip)

**Comment by tandraschko** [comment]
did this work in 9.0?
for me it looks like ListDataModel from JSF isnt serializeable and the default constructor always sets null as wrapped data, so we cant fix it anway.

**Comment by Rapster** [comment]
That's why I asked if it was a regression... But what bugs me is that, it only happens when selection is enabled...

**Comment by tandraschko** [comment]
im not sure either
the NPE just happens in Mojarra, works fine in MyFaces

maybe Mojarra handles serialization bit weird/different as even if i make the DataModel transient, it seems that the same NPE still occurs
but actually it should be recreated in the getter

**Comment by tandraschko** [comment]
nevertheless, i dont think the case is valid as ListDataModel is not seriazable and should not be used like this

**Comment by melloware** [comment]
It does not look like a regression here is an 8.0 version of it that also fails after sorting.
[pf-6913.zip](https://github.com/primefaces/primefaces/files/5913246/pf-6913.zip)

Should we remove support for ListDataModel and throw an error instead?

**Comment by tandraschko** [comment]
It is supported in requestscoped propably

**Comment by melloware** [comment]
@robtha can you confirm this was not working previously in older PrimeFaces please?  I can't get it to work on 8.0 so I have a feeling this use case never worked.

**Comment by tandraschko** [comment]
we cant fix it anyway, we would need to add a new own model
i think the best way is to clarify this with the JSF specs

**Comment by robtha** [comment]
@melloware yes, I can confirm that works with primefaces 8.0.4. We are currently preparing our application for PF10, that's why I ran into this issue.

**Comment by tandraschko** [comment]
we can reopen but i dont have time to work on it.

**Comment by Rapster** [comment]
UIData supports DataModel for a while now, I don't think it's about Serialization here... I may have a clue, when selection is enabled model is set to null when filteredValue is set, remove it and that should work, see SelectionFeature. I'll try to have a closer look. Being said, in 8.0 the problem is different, when I sort first time nothing shows...

**Comment by robtha** [comment]
From my point of view I can live with the workaround using a List instead of a ListDataModel. Most of our views are using a LazyDataModel and we have not discovered any problems so far.
For me the question is whether the cause of the NPE is really limited to the use of ListDataModel only.

**Comment by Rapster** [comment]
Yeah it sounds like there is more than just one problem here... Did you try the same scenario with LazyDataModel?

**Comment by robtha** [comment]
@Rapster no error using LazyDataModel. But sorting on LazyDataModel does not invoke SortFeature.sort (where the NPE occurs)

**Comment by Rapster** [comment]
Yup but should invoke SelectionFeature (if still enabled of course) and so far, my concern is that it happens because of SelectionFeature, but if you're saying it it doesn't... Then... I'm more confused 😅 

Oh but yeah I see what you mean by SortFeature not being called, indeed as it is not called, problem doesn't occur...

**Comment by robtha** [comment]
After some testing I discovered that the NPE on sorting only occurs if filtering is enabled for
a column - it does not need to be the same column used for sorting.

After some debugging I identified the problem in method SelectionFeature.decode.

If the table has filtered values (it even has values if no filter entered) the DataTables value is
set to null and restored with 'originalValue' after decodeSelection.
This way the ValueExpression of the property is lost and replaced by a DataModel instance which seems
to lead into a problem with the state of the component as in further sort requests getValue() returns
an ListTableModel with a null list.

I solved the problem in SelectionFeature by replacing get/setValue() with get/setDataModel().
Maybe you might take a look on the applied patch.

An other circumstance that makes it difficult to reproduce the problem is that DataTable Features are
not processed in a predictable order. As Features are stored in a HashMap, the order might change 
between application restarts - using a LinkedHashMap would help.
[sort_selectionfeature.patch.txt](https://github.com/primefaces/primefaces/files/5931398/sort_selectionfeature.patch.txt)

**Comment by Rapster** [comment]
> If the table has filtered values (it even has values if no filter entered) the DataTables value is
set to null and restored with 'originalValue' after decodeSelection.
This way the ValueExpression of the property is lost and replaced by a DataModel instance which seems
to lead into a problem with the state of the component as in further sort requests getValue() returns
an ListTableModel with a null list.

Yeah that's what I feared https://github.com/primefaces/primefaces/issues/6913#issuecomment-772534362

> An other circumstance that makes it difficult to reproduce the problem is that DataTable Features are
not processed in a predictable order. As Features are stored in a HashMap, the order might change
between application restarts - using a LinkedHashMap would help.

True that's something I actually thought about, but since no issue was raised regarding this I thought maybe I was going the wrong way...

I'll try to have a look today

**Comment by robtha** [comment]
@Rapster As the problem seems somehow related to filtervalues and getFilteredValue always returns a value, even if no filter value entered, I did some investigation on FilterFeature.

Finally applied some minor changes so filteredValue is only set if there is actually a filter value applied.
Personally I found it a little bit confusing to get filteredValues if no filter criteria entered.
And it too solves the NPE with sorting and enabled selection.

I successfully have tested this patch with primefaces showcase samples.

[FilterFeature.patch.txt](https://github.com/primefaces/primefaces/files/5936813/FilterFeature.patch.txt)
